default/python/universe_generation/specials.py

```diff
@@ -1,77 +1,178 @@
 import random
+from collections import defaultdict
+
 import freeorion as fo
 import statistics
 import universe_tables
+from timers import Timer
+from galaxy_topology import get_systems_within_jumps
+
+REPEAT_RATE = {1: 0.08, 2: 0.05, 3: 0.01, 4: 0.00}
+
+def calc_num_placed(objs):
+    """Return a list of number placed at each obj"""
+    return [1 if random.random() > REPEAT_RATE[1] else
+            2 if random.random() > REPEAT_RATE[2] else
+            3 if random.random() > REPEAT_RATE[3] else 1
+            for __ in objs]
+
+
+def place_special_fast(specials, obj, specials_timer, per_special_timer):
+    """ Place at least single special.
+    Return the number of specials placed.
+    """
+    num_placed = 0
+    # Calculate the conditional probabilities that each special is
+    # placed here given that a special will be placed here.
+    probs = [fo.special_spawn_rate(sp) for sp in specials]
+    total_prob = sum(probs)
+    thresholds = [x/total_prob for x in probs]
```

Comment:
use explicit float conversion in division and make sure that you do not divide by zero.